### PR TITLE
*: add top sql agent service definition

### PR DESCRIPTION
**Pull request by dragonly**
Add Top SQL agent service and message definitions.
This is used to send Top SQL records to an agent for persistent storage.

### CODE REVIEW

File: proto/topsql_agent.proto [hunk 1]
```diff
@@ -0,0 +1,35 @@
+syntax = "proto2";
```
Comment (crazycs520) [inline proto/topsql_agent.proto:1]: proto3 ？

Comment (dragonly) [reply proto/topsql_agent.proto]: done

**Review by crazycs520 — COMMENTED** [review]
/lgtm

File: proto/topsql_agent.proto [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+service TopSQLAgent {
+  rpc CollectTiDB(stream CPUTimeRequestTiDB) returns (stream Empty) {}
+  rpc CollectTiKV(stream CPUTimeRequestTiKV) returns (stream Empty) {}
+}
+
+message Empty {}
+
+message CPUTimeRequestTiDB {
+  repeated uint64 timestamp_list = 1;  // timestamp in second
+  repeated uint32 cpu_time_ms_list = 2;  // range is [0, 1000]
```
Comment (breeswish) [inline proto/topsql_agent.proto:23]: maybe larger than 1000, when SQL is executed concurrently.

File: proto/topsql_agent.proto [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  string sql_digest = 3;
+  string normalized_sql = 4;  // SQL string with sensitive fields trimmed, potentially > 1M
+  string plan_digest = 5;
+  string normalized_plan = 6;  // potentially > 1M
+}
+
+message CPUTimeRequestTiKV {
```
Comment (breeswish) [inline proto/topsql_agent.proto:30]: This should be placed inside kvproto.

File: proto/topsql_agent.proto [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+option java_package = "com.pingcap.tidb.tipb";
+
+import "gogoproto/gogo.proto";
+
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.sizer_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+service TopSQLAgent {
+  rpc CollectTiDB(stream CPUTimeRequestTiDB) returns (stream Empty) {}
```
Comment (breeswish) [inline proto/topsql_agent.proto:15]: the response doesn't need to be streamed. It can be unary.

File: proto/topsql_agent.proto [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  string sql_digest = 3;
+  string normalized_sql = 4;  // SQL string with sensitive fields trimmed, potentially > 1M
+  string plan_digest = 5;
+  string normalized_plan = 6;  // potentially > 1M
+}
+
+message CPUTimeRequestTiKV {
+  repeated uint64 timestamp_list = 1;  // timestamp in second
+  repeated uint32 cpu_time_ms_list = 2;  // range [0, 1000]
+  string sql_digest = 3;
```
Comment (breeswish) [inline proto/topsql_agent.proto:33]: TiKV doesn't know about sql_digest or plan_digest.

**Review by breeswish — DISMISSED** [review]

**Comment by breeswish** [comment]
CI failed. The rest LGTM.

**Review by breeswish — APPROVED** [review]

**Review by breeswish — APPROVED** [review]

File: proto/topsql_agent.proto [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import "google/protobuf/empty.proto";
+
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.sizer_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+// TopSQLAgent is the persistent agent service for TopSQL records
+service TopSQLAgent {
+  // CollectCPUTime is called periodically (e.g. per minute) to save the in-memory TopSQL records
+  rpc CollectCPUTime(stream CPUTimeRequestTiDB) returns (google.protobuf.Empty) {}
```
Comment (breeswish) [inline proto/topsql_agent.proto:18]:
```suggestion
  rpc CollectCPUTime(stream CollectCPUTimeRequest) returns (CollectCPUTimeResponse) {}
```

**Review by breeswish — DISMISSED** [review]

**Comment by breeswish** [comment]
Seems that you need to re-generate the code.

File: proto/topsql_agent.proto [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  rpc CollectCPUTime(stream CollectCPUTimeRequest) returns (CollectCPUTimeResponse) {}
+}
+
+message CollectCPUTimeRequest {
+  string sql_digest = 1;
+  string plan_digest = 2;
+  repeated uint64 timestamp_list = 10;  // timestamp in second
+  repeated uint32 cpu_time_ms_list = 11;  // this can be greater than 1000 when counting concurrent running SQL queries
+  string normalized_sql = 20;  // SQL string with sensitive fields trimmed, potentially > 1M
+  string normalized_plan = 21;  // potentially > 1M
```
Comment (crazycs520) [inline proto/topsql_agent.proto:26]: Should add `bool is_internal = 22` field? @breeswish PTAL

Comment (dragonly) [reply proto/topsql_agent.proto:26]: done

**Review by crazycs520 — COMMENTED** [review]
LGTM

**Review by breeswish — APPROVED** [review]

**Review by breeswish — APPROVED** [review]

**Pull request merged**